### PR TITLE
SWITCH to circleci over travisCI && CHANGE lerna bootstrap procedure:

### PR DESCRIPTION
**Pull request by ndelangen**
Issue:
- TravisCI is taking a long time to build 
- Fixes: #1127
- Bootstrapping takes a long time
- A lot of modules are duplicated during bootstrap
  Because of this modules are not singletons during development.
  This leads to a list of problems, like duplicate react, no channel in `lib/addons`..
- Storyshots testing is disabled in cra-kitchen-sink

## What I did
- Configure CircleCI (in parallel)
- Remove TravisCI
- Use yarn for installation of packages
- Hoist all dependencies to root using `--hoist`
- Hoist known packages to root using a custom script
- Test storyshots in cra-kitchen-sink

## How to test
- This branch should be build with CircleCI
- Doing a clean install and bootstrap on this branch should be successful
- After bootstrap unit tests should pass
- After bootstrap starting `example/cra-kitchen-sink` should be successful
- You should feel a feeling of relief and inner happiness knowing how much faster the CI is.

### CODE REVIEW

File: examples/cra-kitchen-sink/.storybook/config.js [hunk 1]
```diff
@@ -20,3 +20,5 @@ function loadStories() {
 }
 
 configure(loadStories, module);
+
+export { getStorybook };
```
Comment (tmeasday) [inline examples/cra-kitchen-sink/.storybook/config.js]: What's going on with this change?

**Review by tmeasday — CHANGES_REQUESTED** [review]
Can we separate these? The circle script seems *awesome*, but it seems an orthogonal concern to "true hoisting"+testing `cra-kitchen-sink`

**Comment by ndelangen** [comment]
No we cannot split this, TravisCI fails and I just don't see a good reason to spend time on fixing it.
https://travis-ci.org/storybooks/storybook/builds/255133024

What's your reason for splitting? is it too big to review? Is there any other way to get this approved? I'm interested in doing a demo video anyway, would that help?

File: examples/cra-kitchen-sink/.storybook/config.js (hunk 1 above)
Comment (ndelangen) [inline examples/cra-kitchen-sink/.storybook/config.js]: getStorybook returns a hash-table of stories.

If the config would export this, we can make storyshots framework agnostic 🎉

Comment (tmeasday) [inline examples/cra-kitchen-sink/.storybook/config.js]: That seems great, although can you explain a bit more why? Should this be part of this PR? It sounds like a seperate thing

**Comment by tmeasday** [comment]
Could you not first do the circle changes and then a second pr targeting the first with the cra-ks changes. 

Yes, I'll admit I have a lot of trouble properly reviewing these big PRs. There are some important subtleties around this symlinking stuff that it is hard enough to keep straight in my head without also needing to figure out if any given change is relevant to the fix. 

Small PRs=non-linearly better imo

File: addons/storyshots/src/index.js [hunk 2]
```diff
@@ -29,36 +27,40 @@ export default function testStorySnapshots(options = {}) {
...
     const configDirPath = path.resolve(options.configPath || '.storybook');
     configPath = path.join(configDirPath, 'config.js');
 
-    const babelConfig = loadBabelConfig(configDirPath);
-    const content = babel.transformFileSync(configPath, babelConfig).code;
-    const contextOpts = {
-      filename: configPath,
-      dirname: configDirPath,
-    };
+    // const babelConfig = loadBabelConfig(configDirPath);
```
Comment (shilman) [inline addons/storyshots/src/index.js]: ???

File: examples/cra-kitchen-sink/.storybook/config.js (hunk 1 above)
Comment (shilman) [inline examples/cra-kitchen-sink/.storybook/config.js]: how does exporting `getStorybook` help exactly? i'm already familiar with the function.

File: examples/react-native-vanilla/storybook/storybook.js [hunk 3]
```diff
@@ -19,4 +19,5 @@ setTimeout(
 );
 
 AppRegistry.registerComponent('ReactNativeVanilla', () => StorybookUI);
-export default StorybookUI;
+
+export { StorybookUI as default, getStorybook };
```
Comment (shilman) [inline examples/react-native-vanilla/storybook/storybook.js]: same question here

File: package.json [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
   "scripts": {
-    "bootstrap": "lerna bootstrap",
-    "bootstrap:docs": "cd docs && npmc install",
-    "bootstrap:test-cra": "npm run build-packs && lerna exec --scope test-cra -- npmc install",
-    "bootstrap:react-native-vanilla": "lerna exec --scope react-native-vanilla -- npmc install",
+    "bootstrap": "lerna bootstrap --concurrency 1 --npm-client=\"yarn\" --hoist && node ./scripts/hoist-internals.js",
+    "bootstrap:docs": "cd docs && yarn install",
+    "bootstrap:react-native-vanilla": "lerna exec --scope react-native-vanilla -- yarn install",
+    "bootstrap:test-cra": "npm run build-packs && lerna exec --scope test-cra -- yarn install",
+    "build": "npm run bootstrap:docs && npm run docs:build",
```
Comment (shilman) [inline package.json]: why do we need this? `bootstrap:docs` alone should be fine?

File: package.json [hunk 5]
```diff
@@ -25,15 +27,14 @@
     "lint:js": "eslint . --cache --cache-location=.cache/eslint --ext .js,.jsx,.json",
     "lint:md": "remark .",
     "publish": "lerna publish",
+    "start": "serve ./docs/out --single",
```
Comment (shilman) [inline package.json]: add `serve` as dev dependency? i think you are mixing in some of your docsv2 stuff in here?

File: scripts/hoist-internals.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            item.code === 0 ? Promise.resolve(locations[index]) : Promise.reject(item)
+        )
+    )
+  );
+
+task
+  .then(() => {
+    shell.echo(chalk.green('COMPLETE'), chalk.gray('=> Hoisting internal packages'));
+  })
+  .catch(() => {
```
Comment (shilman) [inline scripts/hoist-internals.js]: print error

**Review by shilman — CHANGES_REQUESTED** [review]
commented but did not test yet

File: addons/storyshots/src/index.js (hunk 2 above)
Comment (ndelangen) [inline addons/storyshots/src/index.js]: Well correct me if I'm wrong: jest can do babel transpilation, why are we doing it?
I don't understand why we're using require.requireActual either

File: package.json (hunk 4 above)
Comment (ndelangen) [inline package.json]: I don't think I'm calling this script anywhere, will remove.

File: package.json (hunk 5 above)
Comment (ndelangen) [inline package.json]: will remove

File: addons/storyshots/src/index.js (hunk 2 above)
Comment (shilman) [inline addons/storyshots/src/index.js]: if we're not using it, why comment it out? why not just delete it?

File: package.json (hunk 4 above)
Comment (shilman) [inline package.json]: i think you can get rid of `docs:build` too

**Review by shilman — APPROVED** [review]

**Comment by ndelangen** [comment]
🎉

File: lib/addons/src/index.js [hunk 7]
```diff
@@ -2,7 +2,7 @@ export class AddonStore {
   constructor() {
     this.loaders = {};
     this.panels = {};
-    this.channel = null;
+    this.channel = { on() {}, emit() {} };
```
Comment (tmeasday) [inline lib/addons/src/index.js]: Hmm, I think this change is going to just make it harder to debug problems with multiple loading. Why did you add it @ndelangen ?

Comment (shilman) [inline lib/addons/src/index.js]: Great point. Let's take this out unless it's necessary!

Comment (ndelangen) [inline lib/addons/src/index.js]: This was giving us so much grief, I thought we could do without?

To the best of my abilities, everything actually works, there's no multiple loading going on.

It should never occur to not have set the channel.

Having said that, I think it's just really wrong to have a setup where side effects have such a major impact on whether a test can be run.

The channels is completely unneeded to run the snapshots for example, yet they have caused a lot of pain for a lot of users.

I just want people to be able to use and develop this. I think this code facilitates that better then without.

Comment (shilman) [inline lib/addons/src/index.js]: @ndelangen completely disagree. if this doesn't get overwritten it will cause an error to fail silently! having been bit by this bug before, it will be even harder to debug when there is a problem!

Comment (tmeasday) [inline lib/addons/src/index.js]: I have to agree with @shilman here, it's better that our or even a user's tests fail than they can't actually run storybook or an addon in the first place and can't figure out why (it's hard enough with the channel undefined error in the first place).

Comment (tmeasday) [inline lib/addons/src/index.js]: If we want to have a way to "ignore" the error, maybe that's a feature worth adding, if it's sometimes impossible to avoid the multi-loading problem. Something like `initStoryshots({ nullChannel: true})`?
